src/common/enumiterator.h

```diff
@@ -0,0 +1,155 @@
+
+#ifndef ENUMITERATOR_H
+#define ENUMITERATOR_H
+
+#include <cassert> // assert
+#include <cstddef> // ptrdiff_t, size_t
+#include <iterator> // forward_iterator_tag
+#include <memory> // unique_ptr
+#include <limits> // numeric_limits
+#include <type_traits> // {common,underlying}_type, is_enum, add_{pointer,lvalue_reference,const}, conditional
+#include <utility> // swap
+
+
+
+/**
+ * Iterator for an enum of type @c T that starts with an item @c start
+ * and ends *after* the item @c stop. Models ForwardIterator concept.
+ *
+ * @warning Following iterator DOES NOT WORK for non-contiguous enumerations
+ *          like "enum class E { A = 100, B = 200, C == 300 }".
+ *
+ * @see http://en.cppreference.com/w/cpp/concept/ForwardIterator
+ */
+template <class T, T start, T stop>
+class EnumIterator
+{
+
+    using raw_type = std::underlying_type_t<T>;  // value_type = T
+
+    /**
+     * If [start, stop] spawns [min, max], then past-the-end item is the MAX
+     * of size_t or ssize_t depending on the sign of the value_type.
+     *
+     * Value of stop equal to selected MAX is not allowed!
+     *
+     * @{ */
+    using stored_type =
+            std::common_type_t<
+                raw_type
+              , std::conditional_t<
+                    std::is_signed<raw_type>::value
+                  , std::ptrdiff_t // FIXME: ssize_t?
```

Comment:
Why do we have fixme here? Can this be fixed before PR is merged?

Reply:
I would use `ssize_t`, but that type is not a standard one. Maybe one day it will go into standard, thus FIXME. `ptrdiff_t` works fine here.